slam-libraries/viam-cartographer/src/slam_service/slam_service.cc

```diff
@@ -329,6 +325,110 @@ void SLAMServiceImpl::SetUpMapBuilder() {
 }
 
 std::string SLAMServiceImpl::GetLatestJpegMapString(bool add_pose_marker) {
+    
+    std::unique_ptr<cartographer::io::PaintSubmapSlicesResult> painted_slices = nullptr;
+    try {
+        painted_slices = std::make_unique<cartographer::io::PaintSubmapSlicesResult>(GetLatestPaintedMapSlices());
+    } catch (std::exception &e) {
+        // jpeg will be empty which captures the error
+        return "";
+    }
+    if (add_pose_marker) {
+        PaintMarker(painted_slices.get());
+    }
+
+    auto image = io::Image(std::move(painted_slices->surface));
+    return image.WriteJpegToString(jpegQuality);
+}
+
+void SLAMServiceImpl::GetLatestPointCloudMapString(std::string &pointcloud) {
+
+    std::unique_ptr<cartographer::io::PaintSubmapSlicesResult> painted_slices = nullptr;
+    try {
+        painted_slices = std::make_unique<cartographer::io::PaintSubmapSlicesResult>(GetLatestPaintedMapSlices());
+    } catch (std::exception &e) {
+        // pointcloud will be empty which captures the error
+        return;
+    }
+
+    auto painted_surface = painted_slices->surface.get();
+    int width = cairo_image_surface_get_width(painted_surface);
+    int height = cairo_image_surface_get_height(painted_surface);
+    auto data = cairo_image_surface_get_data(painted_surface);
+
+    // total number of bytes in image (4 bytes per pixel)
+    int size_data = width * height * 4;
+
+    std::vector<unsigned char> data_vect(data, data + size_data);
+
+    int num_points = 0;
+
+    // scaling factor to ensure number of pixels is within required size
+    // do not set to 0
+    float scaleFactor = .5;
+
+    // Reduce resolution based off number of pixels. Output is pixels to skip
+    // Ideally should be between 5 and 15, but depends on the resolution of the
+    // original image
+    int scale = size_data / maximumGRPCByteLimit * scaleFactor;
```

Comment:
[q] does this imply we will always send a max size image? Is there a minimum/maximum scale we want to limit this to be so that if we have a tiny map the jpeg might only be 2MB and not always 32MB?

Reply:
The scale/sampling factor is letting us control the size of the image as well. Also these parameters are just giving us a sampling value to work with to ensure we are within the limits, and is not meant to the PCD exactly at 32MB